Add Piwik tracking

KZeni opened the pull request:
Why limit this plugin to offering Google Analytics? I've added Piwik tracking code, settings, etc. for offering HTTP-based tracking and JavaScript-based tracking.

The existing Google Analytics tracking code continues to work as expected, but now users can optionally provide Piwik tracking information to have it tracked there instead/as well.

I would love to see this officially adopted as I have a client in China where Google is blocked so Piwik is the next best thing, and I'd like to continue using this plugin considering how great it is already.

As an aside, I've also posted about this on the WordPress.org forum at https://wordpress.org/support/topic/proposed-patch-add-piwik-tracking-support/

ronalfy commented:
Thanks for the PR. The code looks legit, so I’ll give it a spin shortly.
What’s your .org profile? I’d like to give you contribution props.
On Thu, Jan 4, 2018 at 23:29 Kurt Zenisek <notifications@github.com> wrote:

> Why limit this plugin to offering Google Analytics? I've added Piwik
> tracking code, settings, etc. for offering HTTP-based tracking and
> JavaScript-based tracking.
>
> The existing Google Analytics tracking code continues to work as expected,
> but now users can optionally provide Piwik tracking information to have it
> tracked there instead/as well. Both have been tested and are confirmed to
> be working.
>
> I would love to see this officially adopted as I have a client in China
> where Google is blocked so Piwik is the next best thing, and I'd like to
> continue using this plugin considering how great it is already.
> ------------------------------
> You can view, comment on, or merge this pull request online at:
>
>
> https://github.com/ronalfy/wordpress-gravity-forms-event-tracking/pull/28
> Commit Summary
>
>    - Add piwik code
>    - Added piwik settings
>    - Added piwik HTTP API
>    - Added piwik code
>    - Added piwik code
>
> File Changes
>
>    - *M* gravity-forms-event-tracking.php
>    <https://github.com/ronalfy/wordpress-gravity-forms-event-tracking/pull/28/files#diff-0>
>    (21)
>    - *M* includes/GFGAET_Pagination.php
>    <https://github.com/ronalfy/wordpress-gravity-forms-event-tracking/pull/28/files#diff-1>
>    (84)
>    - *A* includes/GFGAET_Piwik_HTTP_API.php
>    <https://github.com/ronalfy/wordpress-gravity-forms-event-tracking/pull/28/files#diff-2>
>    (154)
>    - *M* includes/GFGAET_Submission_Feeds.php
>    <https://github.com/ronalfy/wordpress-gravity-forms-event-tracking/pull/28/files#diff-3>
>    (121)
>    - *M* includes/GFGAET_UA.php
>    <https://github.com/ronalfy/wordpress-gravity-forms-event-tracking/pull/28/files#diff-4>
>    (68)
>
> Patch Links:
>
>    -
>    https://github.com/ronalfy/wordpress-gravity-forms-event-tracking/pull/28.patch
>    -
>    https://github.com/ronalfy/wordpress-gravity-forms-event-tracking/pull/28.diff
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/ronalfy/wordpress-gravity-forms-event-tracking/pull/28>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAm2adXrB5SY3Udjwi4gzRfZwDs3x89Yks5tHbNBgaJpZM4RUDF_>
> .
>

ronalfy commented on includes/GFGAET_Piwik_HTTP_API.php:
```diff
@@ -0,0 +1,154 @@
+
+		$ga_options = get_option( 'gravityformsaddon_GFGAET_UA_settings', false );
+		if ( ! isset( $ga_options[ 'gravity_forms_event_tracking_piwik_url' ] ) ) {
+			$this->endpoint = 'http://www.example.com/piwik/piwik.php'; // Default Tracking HTTP API Endpoint
```
The PR looks good, but I'm concerned about this section and if data is always assumed to be sent to this endpoint. I think a check is needed and if people don't fill out the Piwik data, then there should be a check that ignores sending.

KZeni replied:
Ah, I did take a shortcut there which threw away the data rather than checking if the required endpoint and site id are provided. I also have some spots where I have it still checking if the Google info was provided rather than the Piwik info was provided in some spots.

I'll see about getting that updated. Otherwise, feel free to edit my PR as you see fit!

ronalfy commented:
I don’t want to break anything. I have very little insight into how Piwik
can be used, so I’m trusting you :)
On Mon, Jan 8, 2018 at 18:51 Kurt Zenisek <notifications@github.com> wrote:

> *@KZeni* commented on this pull request.
> ------------------------------
>
> In includes/GFGAET_Piwik_HTTP_API.php
> <https://github.com/ronalfy/wordpress-gravity-forms-event-tracking/pull/28#discussion_r160295803>
> :
>
> > +	private $rec = 1; // Required for tracking
> +	private $apiv = 1; // API Version (currently v1)
> +	private $rand = ''; // Random value to help avoid the tracking request being cached by the browser or a proxy
> +	private $action_name = ''; // Event category
> +	private $e_c = ''; // Event category
> +	private $e_a = ''; // Event action
> +	private $e_n = ''; // Event Label
> +	private $e_v = ''; // Event Value
> +	private $url = ''; // Full URL
> +
> +	public function init() {
> +		$this->_id = $this->create_client_id();
> +
> +		$ga_options = get_option( 'gravityformsaddon_GFGAET_UA_settings', false );
> +		if ( ! isset( $ga_options[ 'gravity_forms_event_tracking_piwik_url' ] ) ) {
> +			$this->endpoint = 'http://www.example.com/piwik/piwik.php'; // Default Tracking HTTP API Endpoint
>
> Ah, I did take a shortcut there which threw away the data rather than
> checking if the required endpoint and site id are provided. I also have
> some spots where I have it still checking if the Google info was provided
> rather than the Piwik info was provided in some spots.
>
> I'll see about getting that updated. Otherwise, feel free to edit my PR as
> you see fit!
>
> —
> You are receiving this because you commented.
>
>
> Reply to this email directly, view it on GitHub
> <https://github.com/ronalfy/wordpress-gravity-forms-event-tracking/pull/28#discussion_r160295803>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAm2ae28rraUeXZDkkoJzeAQqeo3NYMRks5tIrgPgaJpZM4RUDF_>
> .
>

KZeni commented:
Alright. I've updated the code to utilize its own checks to see if Matomo has been configured.

Also, Piwik has just been renamed to Matomo ([see the announcement here](https://matomo.org/blog/2018/01/piwik-is-now-matomo/)) so I've updated the naming accordingly.
